### PR TITLE
Add documentation for `Bimonad`

### PR DESCRIPTION
**Pull request by gatear**
The first draft for `Bimonad` docs as required in #1801
Couldn't find a better example ..  i'm open to suggestions.

### CODE REVIEW

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (armanbilge) [inline docs/src/main/mdoc/typeclasses/bimonad.md:10]: I wonder if this is confusing—even though it doesn't introduce new methods, as you note below it does have additional laws which do affect (restrict, actually) it's behavior relative to it's parent traits. Maybe prefer the word method here?

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (armanbilge) [inline docs/src/main/mdoc/typeclasses/bimonad.md:11]:
```suggestion
different from other `Bi` typeclasses like `Bifunctor`, `Bifoldable` or `Bitraverse` where the prefix describes a
```

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (armanbilge) [inline docs/src/main/mdoc/typeclasses/bimonad.md:12]: "dual monad" is another one that I worry is confusing. For one, `Comonad` is not a `Monad`. Actually it is the dual of `Monad`!
https://typelevel.org/cats/typeclasses/comonad.html

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (armanbilge) [inline docs/src/main/mdoc/typeclasses/bimonad.md:20]: Just a suggestion, to show the relationship to both parents.
```suggestion
Eval is a lawful `Bimonad` so you can chain computations (like a `Monad`) and `extract` the result at the end (like a `Comonad`).
```

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (armanbilge) [inline docs/src/main/mdoc/typeclasses/bimonad.md:28]: I think what you are trying to show here is that `pure` is implemented as `now` and `extract` as `value`? One alternative would be to define `Bimonad[Eval]` here just for demonstrative purposes, WDYT?

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (armanbilge) [inline docs/src/main/mdoc/typeclasses/bimonad.md:46]: I'm not an `mdoc` expert but I'm pretty sure it will automatically render the result when it runs :) I think you can preview this by running `docs/mdoc --watch` in sbt but not 100% sure.
```suggestion
```

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (armanbilge) [inline docs/src/main/mdoc/typeclasses/bimonad.md:54]:
```suggestion
```

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (armanbilge) [inline docs/src/main/mdoc/typeclasses/bimonad.md:49]:
```suggestion
`Function0` and `NonEmptyList` are also lawful bimonads so the following calls are also valid:
```

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (gatear) [reply docs/src/main/mdoc/typeclasses/bimonad.md:10]: Yep i agree. I'll use `without introducing new methods`

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (gatear) [reply docs/src/main/mdoc/typeclasses/bimonad.md:12]: The `DualMonad`\\`SelfDualMonad` thing i picked from old issues like #1297

I'll change it to:
```
The `Bimonad` is a `F[_]` thus the `Bi` prefix has a different meaning here: it's both a `Monad` and a `Comonad`.
```

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (gatear) [reply docs/src/main/mdoc/typeclasses/bimonad.md:28]: > I think what you are trying to show here is that `pure` is implemented as `now` and `extract` as `value`?
Yes

> One alternative would be to define Bimonad[Eval] here just for demonstrative purposes, WDYT?

Something like that ?
```scala
val evalBimonad = Bimonad[Eval]

evalBimonad.pure(true).extract === Eval.now(true).value
```

Comment (armanbilge) [reply docs/src/main/mdoc/typeclasses/bimonad.md:28]: I was thinking like how they define `optionMonad` in the `Monad` docs:
http://typelevel.org/cats/typeclasses/monad.html

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (gatear) [reply docs/src/main/mdoc/typeclasses/bimonad.md]: Yep it was helpful. Thanks!

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (johnynek) [inline docs/src/main/mdoc/typeclasses/bimonad.md:36]: `Eval[_]` is a bit of a bad choice here because the Comonad is not very safe. Calling `.value` on an Eval must be done with some care (it needs to be at O(1) stack depth or you lose the stack safety.  If you are calling this as part of a recursive call it often blows up).

`Id[_]` is also a Comonad, but a bit of a boring one. I wonder if `NonEmptyList[_]` would be a better example since it is non-trivial but also not somewhat unsafe in this way.

Comment (gatear) [reply docs/src/main/mdoc/typeclasses/bimonad.md:36]: Hm ok, I was hesitating between `Eval[_]` and `NonEmptyList[_]` for the main example.
I'll just use `NonEmptyList[_]`.

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (armanbilge) [inline docs/src/main/mdoc/typeclasses/bimonad.md:47]: Actually we can directly call `flatMap` and `coflatMap` directly on the `NonEmptyList` :)

Comment (gatear) [reply docs/src/main/mdoc/typeclasses/bimonad.md:47]: I was hopping this shows `flatMap` and `coflatMap` as originating from `monad` and `comonad`. 
Don't you think it's a good idea to delegate to the `monad` and `comonad` variables ?

Comment (armanbilge) [reply docs/src/main/mdoc/typeclasses/bimonad.md:47]: Yes, I do see that now. But then isn't it confusing, because `pure` and `extract` are also defined in `monad` and `comonad`, so we should delegate those as well?

Comment (armanbilge) [reply docs/src/main/mdoc/typeclasses/bimonad.md:47]: > I thought using one and head in the implementation will make this more obvious

I agree, I think this is the right way to do it. So, we can just delegate `flatMap` and `coflatMap` to the list without `monad`/`comonad` and maybe leave `tailRecM` as an exercise for the reader 😉

Anyway just my opinion, not really sure what the best thing to do here is. But I think separating the implementations of `pure` and `extract` makes it seem like they are inconsistent or different from the definitions in `monad` or `comonad`.

Comment (gatear) [reply docs/src/main/mdoc/typeclasses/bimonad.md]: So the current `tailRecM` implementation for `NonEmptyList[_]` doesn't look trivial .. I mean I fear Its understanding detours too much and I agree it's a good idea to leave it as an exercise 👍 

> Anyway just my opinion, not really sure what the best thing to do here is. But I think separating the implementations of pure and extract makes it seem like they are inconsistent or different from the definitions in monad or comonad.

Yes, I will remove `monad` and `comonad`.

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (armanbilge) [inline docs/src/main/mdoc/typeclasses/bimonad.md:29]:
```suggestion
Here is a possible implementation:
```

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (armanbilge) [inline docs/src/main/mdoc/typeclasses/bimonad.md:38]: Maybe a good place to point out the `Bimonad` law about the relationship between `pure` and `extract`? :)

Comment (gatear) [reply docs/src/main/mdoc/typeclasses/bimonad.md:38]:
```scala
    //nelBimonad needs to respect: `nelBimonad.extract(nelBimonad.pure(a)) <-> a`
``` 
Something like this ?

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (armanbilge) [inline docs/src/main/mdoc/typeclasses/bimonad.md:38]: Just a personal style nitpick but I prefer a space here between `//` and the comment 😅 

```suggestion
    // in order to have a lawful bimonad `pure` and `extract` need to respect: `nelBimonad.extract(nelBimonad.pure(a)) <-> a`
```

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (armanbilge) [inline docs/src/main/mdoc/typeclasses/bimonad.md:24]: I think this is important enough to include at the beginning ...

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (armanbilge) [inline docs/src/main/mdoc/typeclasses/bimonad.md:13]: ... like this:
```suggestion
Keep in mind `Bimonad` has its own added laws so something that is both monadic
and comonadic may not necessarily be a lawful `Bimonad`.

```

**Review by armanbilge — COMMENTED** [review]
Thanks for iterating on all the feedback! Looking good, just a couple tiny things :)

**Review by armanbilge — APPROVED** [review]
Thanks for your contribution! :)

**Comment by gatear** [comment]
@armanbilge is this ready to merge or do i need 2 reviews ? I want to start working on another doc for `Bifoldable` after this one is merged

**Comment by armanbilge** [comment]
Another pair of eyes is always good, maybe @johnynek can take another look at the next opportunity :)

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (johnynek) [inline docs/src/main/mdoc/typeclasses/bimonad.md:18]: I think this should be `def f[T[_]: Monad: Comonad, S](fa: T[S]): S`

I think the comma here is making`Comonad` just a type parameter.

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (johnynek) [inline docs/src/main/mdoc/typeclasses/bimonad.md:22]: nit: we generally don't put a space after the type, so we would change to: `f[T[_]: Bimonad, S]`

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (johnynek) [inline docs/src/main/mdoc/typeclasses/bimonad.md:35]: I would make this a `implicit val` since it doesn't have any args.

File: docs/src/main/mdoc/typeclasses/bimonad.md
Comment (gatear) [reply docs/src/main/mdoc/typeclasses/bimonad.md:18]: Yep my bad ..`:Monad :Comonad` is what I wanted

**Review by johnynek — APPROVED** [review]

**Comment by johnynek** [comment]
Thanks!

**Pull request merged**
